Add live queries blog for 3.0 (#3776)
opensearch-project/project-website: _posts/2025-05-09-query-insights-live-queries.md

---
layout: post
title: "Real-time query monitoring with live queries in OpenSearch 3.0"
authors:
   - chenyang

> date: 2025-05-09

Reviewer: Update publish date: 2025-06-05

    *   `latency`: The current running time, in nanoseconds.
    *   `cpu`: The CPU time consumed so far, in nanoseconds.
    *   `memory`: The amount of heap memory used so far, in bytes.

## Getting started with live queries

Interacting with the Live Queries API is straightforward. The API supports the following operations.

### Basic request

To get a list of currently running queries sorted by latency, send the following request:

```json
GET /_insights/live_queries
```

### Customizing your view using query parameters

You can tailor the Live Queries API response using the following optional parameters:
- `verbose`: Includes detailed query information, such as the query source.
- `nodeId`: Filters results by a comma-separated list of node IDs.
- `sort`: Sorts results by a specific metric, such as latency, CPU usage, or memory usage.
- `size`: Limits the number of query records returned.

### Example: Finding CPU-intensive live queries

To find the top five queries currently consuming the most CPU, you can provide the following query parameters:

```json
GET /_insights/live_queries?verbose=false&sort=cpu&size=5
```

> {% include copy-curl.html %}

Reviewer 2: ```suggestion
```

### Key information returned

For each live query, you'll receive a rich set of data:

*   `timestamp`: The time at which the query task started, in milliseconds since the epoch.
*   `id`: The unique search task ID.
*   `description`: Details about the query, including target indexes, search type, and the query source itself (if `verbose` is true).
*   `node_id`: The ID of the coordinator node on which the query task is running.
*   `measurements`: An object containing performance metrics gathered so far:
    *   `latency`: The current running time, in nanoseconds.
    *   `cpu`: The CPU time consumed so far, in nanoseconds.
    *   `memory`: The amount of heap memory used so far, in bytes.

## Getting started with live queries

Interacting with the Live Queries API is straightforward. The API supports the following operations.

### Basic request

To get a list of currently running queries sorted by latency, send the following request:

```json
GET /_insights/live_queries
```

### Customizing your view using query parameters

You can tailor the Live Queries API response using the following optional parameters:

> - `verbose` -- Includes detailed query information, such as the query source.

Reviewer 3: ```suggestion
- `verbose`: Includes detailed query information, such as the query source.
```

> - `nodeId` -- Filters results by a comma-separated list of node IDs.

Reviewer 3: ```suggestion
- `nodeId`: Filters results by a comma-separated list of node IDs.
```

> - `sort` -- Sorts results by a specific metric, such as latency, CPU usage, or memory usage.

Reviewer 3: ```suggestion
- `sort`: Sorts results by a specific metric, such as latency, CPU usage, or memory usage.
```

> - `size` -- Limits the number of query records returned.

Reviewer 3: ```suggestion
- `size`: Limits the number of query records returned.
```

---
layout: post
title: "Real-Time Query Monitoring with Live Queries in OpenSearch 3.0"
authors:
   - chenyang
date: 2025-05-09
categories:
  - technical-posts
meta_keywords: OpenSearch Query Insights, Live Queries, real-time query monitoring, query performance, OpenSearch 3.0, search analytics, resource monitoring, query debugging
meta_description: Discover the new Live Queries API in OpenSearch Query Insights 3.0, enabling real-time visibility into currently executing search queries to identify and debug performance issues as they happen.
---

> OpenSearch Query Insights has become an important tool for understanding search query performance, offering visibility into how queries execute and consume cluster resources. Building on our commitment to help you identify performance bottlenecks and optimize search operations, we're excited to introduce a powerful new capability in OpenSearch 3.0: **live queries**.

Reviewer 3: ```suggestion
OpenSearch Query Insights has become an important tool for understanding search query performance, offering visibility into how queries execute and consume cluster resources. Building on our commitment to helping you identify performance bottlenecks and optimize search operations, we're excited to introduce a powerful new capability in OpenSearch 3.0: **live queries**.
```

```json
GET /_insights/live_queries
```

By default, the API returns a list of currently executing search queries, sorted by `latency` in descending order.

### Key information returned

For each live query, you'll receive a rich set of data:

*   `timestamp`: The time the query task started, in milliseconds since epoch.
*   `id`: The unique search task ID.
*   `description`: Details about the query, including target indexes, search type, and the query source itself (if `verbose` is true).
*   `node_id`: The ID of the coordinator node on which the query task is running.
*   `measurements`: An object containing performance metrics gathered so far:
    *   `latency`: The current running time, in nanoseconds.
    *   `cpu`: The CPU time consumed so far, in nanoseconds.
    *   `memory`: The amount of heap memory used so far, in bytes.

## Getting started with live queries

Interacting with the Live Queries API is straightforward. The API supports the following operations.

### Basic request

To get a list of currently running queries sorted by latency, send the following request:

```json
GET /_insights/live_queries
```

> ### Customizing your view using query parameters

Reviewer 2: I think it's better to provide a brief overview rather than a parameter table. We can link to the documentation. I'll make a suggestion.

You can tailor the output using several optional query parameters.

| Parameter | Data type | Description                                                                                                |

> | :-------- | :-------- | :--------------------------------------------------------------------------------------------------------- |

Reviewer 3: ```suggestion
```

| `verbose` | Boolean   | Whether to include detailed query information (like the query source) in the output. Default is `true`.    |

> | `nodeId`  | String    | A comma-separated list of node IDs to filter results. If omitted, queries from all nodes are returned.     |

Reviewer 3: ```suggestion
| `nodeId`  | String    | A comma-separated list of node IDs used to filter results. If omitted, queries from all nodes are returned.     |
```

> | `sort`    | String    | The metric to sort results by. Valid values: `latency`, `cpu`, or `memory`. Default is `latency`.          |

Reviewer 3: ```suggestion
| `sort`    | String    | The metric by which to sort results. Valid values: `latency`, `cpu`, or `memory`. Default is `latency`.          |
```

> | `size`    | Integer   | The number of query records to return. Default is `100`.                                                     |

Reviewer 3: ```suggestion                                                     |
```

### Example: Finding CPU-intensive live queries

To find the top 5 queries currently consuming the most CPU, you can provide the following query parameters:

```json
GET /_insights/live_queries?verbose=false&sort=cpu&size=5
```
{% include copy-curl.html %}

### Understanding the response

> Here’s an example of what the response might look like (showing one query for brevity, taken from the [documentation example](https://docs.opensearch.org/docs/latest/observing-your-data/query-insights/live-queries/#example-response)):

Reviewer 3: ```suggestion
Here's an example of what the response might look like (showing one query for brevity, taken from the [documentation example](https://docs.opensearch.org/docs/latest/observing-your-data/query-insights/live-queries/#example-response)):
```

```json
{
  "live_queries" : [
    {
      "timestamp" : 1745359226777,
      "id" : "troGHNGUShqDj3wK_K5ZIw:512",
      "description" : "indices[my-index-*], search_type[QUERY_THEN_FETCH], source[{\"size\":20,\"query\":{\"term\":{\"user.id\":{\"value\":\"userId\",\"boost\":1.0}}}}]",
      "node_id" : "troGHNGUShqDj3wK_K5ZIw",
      "measurements" : {
        "latency" : {
          "number" : 13959364458,
          "count" : 1,
          "aggregationType" : "NONE"
        },
        "memory" : {
          "number" : 3104,
          "count" : 1,
          "aggregationType" : "NONE"
        },
        "cpu" : {
          "number" : 405000,
          "count" : 1,
          "aggregationType" : "NONE"
        }
      }
    }
    // ... other live queries
  ]
}
```

This response provides the following information:

*   The query started at timestamp `1745359226777`.
*   It's running on node `troGHNGUShqDj3wK_K5ZIw`.
*   So far, it has been running for over 13.9 seconds (`latency.number` in nanoseconds).
*   It has consumed `405000` nanoseconds of CPU time and `3104` bytes of memory.

> *   The `description` (provided because `verbose=true` in the request) shows the query targets `my-index-*` and includes the actual query structure.

Reviewer 3: ```suggestion
*   The `description` (provided because `verbose=true` in the request) shows that the query targets `my-index-*` and includes the actual query structure.
```
